### PR TITLE
[Chore] tools: add offline OTA image builder

### PR DESCRIPTION
**Pull request by Bodong-Yang**
## Description

This PR introduces helper script that bundles several OTA images(for several target ECUs) and generates offline OTA image with input images. 

Supported features for this script:
1. bundle multiple OTA images into one offline OTA image,
2. export built offline OTA image as tar archive,
3. prepare specific device into external cache source dev.

The device prepared by this helper script can be used directly as external cache source device recognized by otaproxy(Or user can manually prepare the device and write the built image rootfs to the device).
The built offline OTA image can be or used by full offline OTA helper script to start fully offline OTA.
> **Note**
> The **full offline OTA helper script** is not yet implemented.

For usage document and other information, check the `README.md` under `tools/offline_ota_image_builder`.

## Check list

- [x] design docs/implementation docs are prepared.
- [x] image build mechanism is working.
- [x] image export mechanism is working.
- [x] the built image is valid and VM OTA test passed with built image.

## Documents

1. [OTA cache mechanism design for otaproxy](https://tier4.atlassian.net/l/cp/yzmnPx9T): check **External cache source** section for the specification of external cache source image layout and external cache source dev.

## Related links & tickets

1. [RT4-5484](https://tier4.atlassian.net/browse/RT4-5484)

[RT4-5484]: https://tier4.atlassian.net/browse/RT4-5484?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tools/offline_ota_image_builder/README.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+`tools.offline_ota_image_builder` is a helper package for building offline OTA image from several input OTA images. Built image can be used as external cache source for otaproxy, or used by full offline OTA helper util to trigger fully offline OTA update.
+
+**NOTE**: full offline OTA helper util is not yet implemented.
+
+It provides the following features:
+
+1. build offline OTA image with several input OTA images,
+2. export built image as tar archive,
+3. prepare and export built image onto specific block device as external cache source dev.
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/README.md:11]: `external cache source dev` is source device? (what is `dev`?)

File: tools/offline_ota_image_builder/README.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+Please check documentation at [OTA cache mechanism design for otaproxy](https://tier4.atlassian.net/l/cp/yzmnPx9T) for more details.
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/README.md:13]: This link is not valid for public user. (It is OK if there is note that the link is for the internal developer)

File: tools/offline_ota_image_builder/README.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+## Expected image layout for input OTA images
+
+Please refere to [ota-metadata](https://github.com/tier4/ota-metadata) repo for OTA image layout specification.
+This tool is implemented based on the image layout implemented in [ota-metadata@a711013](https://github.com/tier4/ota-metadata/commit/a711013).
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/README.md:18]: What do you mean by `This tool is implemented based on the image layout`?
Is this means `This tool is implemented based on the image layout as a input OTA image`?

File: tools/offline_ota_image_builder/README.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    ├── <idx> # corresponding to the order in manifest.image_meta
+    │   └── <list of OTA metafiles...>
+    └── ...
+        └── <list of OTA metafiles...>
+
+```
+
+This layout is compatible with external cache source's layout, which means any offline OTA image can natually directly be used as external cache source recognized by otaproxy(but user still needs to prepare the device either by themselves and extract the offline OTA image rootfs onto the prepared device, or prepare the device with this **offline_ota_image_builder** package).
+
+Please check [External cache source](https://tier4.atlassian.net/wiki/spaces/WEB/pages/2813984854/OTA+cache+mechanism+design+for+otaproxy#External-cache-source) section of the doc for more details.
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/README.md:58]: This link is also internal.

File: tools/offline_ota_image_builder/README.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+      "ecu_id": <ecu_id>,
+      "image_version": "<image_version>",
+      "ota_metadata_version": 1,
+    },
+    ...
+  ],
+}
+```
+
+Please check [Build external cache source](https://tier4.atlassian.net/wiki/spaces/WEB/pages/2813984854/OTA+cache+mechanism+design+for+otaproxy#Build-external-cache-source) section for more details.
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/README.md:83]: This link is also internal.

File: tools/offline_ota_image_builder/README.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                        OTA image for <ECU_ID> as tar archive(compressed or uncompressed), this
+                        option can be used multiple times to include multiple images.
+  -o <OUTPUT_PATH>, --output <OUTPUT_PATH>
+                        save the generated image rootfs into tar archive to <OUTPUT_PATH>.
+  -w <DEVICE>, --write-to <DEVICE>
+                        write the image to <DEVICE> and prepare the device as external cache source
+                        device.
+  --confirm-write-to    writing generated to <DEVICE> without inter-active confirmation, only valid
+                        when used with -w option.
+```
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/README.md:173]: It is better to move this block right after `### Build image and export` section to understand the tool usage, and it is better to add an explanation after that.

**Review by obi-t4 — COMMENTED** [review]
I left some comments, please have a check.

File: tools/offline_ota_image_builder/__main__.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                    image_version=_image_version,
+                )
+            )
+            image_files[_ecu_id] = _image_fpath
+        else:
+            logger.warning(f"ignore illegal image pair: {raw_pair}")
+
+    if not image_metas:
+        print("ERR: at least one valid image should be given, abort")
+        sys.exit(-1)
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/__main__.py:74]: positive value should be used in general. this code is referenced as shell return code.

File: tools/offline_ota_image_builder/__main__.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        "-w",
+        "--write-to",
+        help=(
+            "write the image to <DEVICE> and prepare the device as "
+            "external cache source device."
+        ),
+        metavar="<DEVICE>",
+    )
+    parser.add_argument(
+        "--confirm-write-to",
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/__main__.py:143]: confirm-write-to seems to be an option that requires confirmation.
The default is confirm, and I think it is natural if --force is specified not to confirm.

File: tools/offline_ota_image_builder/builder.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    return False
+
+
+@contextmanager
+def _unarchive_image(image_fpath: StrPath, *, workdir: StrPath):
+    _start_time = time.time()
+    with tempfile.TemporaryDirectory(dir=workdir) as unarchive_dir:
+        cmd = f"tar xf {image_fpath} -C {unarchive_dir}"
+        try:
+            subprocess_call(cmd)
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/builder.py:51]: why do you use subprocess, you can use tar python library.

File: tools/offline_ota_image_builder/configs.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    IMAGE_UNARCHIVE_WORKDIR = "images"
+
+    DEFAULT_OTA_METADATA_VERSION = 1
+    OTA_METAFILES_LIST = [
+        "certificate.pem",
+        "metadata.jwt",
+        "dirs.txt",
+        "regulars.txt",
+        "symlinks.txt",
+        "persistents.txt",
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/configs.py:34]: These file names should be parsed from metadata.jwt.

File: tools/offline_ota_image_builder/configs.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    ]
+    OTA_METAFILE_REGULAR = "regulars.txt"
+    OTA_IMAGE_DATA_DIR = "data"
+    OTA_IMAGE_DATA_ZST_DIR = "data.zst"
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/configs.py:38]: This name should be retrieved from metadata.jwt

**Review by obi-t4 — COMMENTED** [review]
I left some additional comments.

File: tools/offline_ota_image_builder/README.md [hunk 12]
```diff
@@ -0,0 +1,173 @@
+# Offline OTA image builder
+
+`tools.offline_ota_image_builder` is a helper package for building offline OTA image from several input OTA images. Built image can be used as external cache source for otaproxy, or used by full offline OTA helper util to trigger fully offline OTA update.
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply tools/offline_ota_image_builder/README.md:3]: Ah I think the `input` word is not needed. 

What about the following updated one:
`tools.offline_ota_image_builder is a helper package for bundling several OTA images into one offline OTA image.`

File: tools/offline_ota_image_builder/README.md (hunk 1 above)
Comment (Bodong-Yang) [reply tools/offline_ota_image_builder/README.md:11]: Ah the word `external cache source` is the terminology from the [OTA cache mechanism design documentation](https://tier4.atlassian.net/l/cp/iSPEDWhT), which means cache source that can be used by otaproxy.
`dev` here means `device`, especially stands for block device. I should not use abbreviation here.

What about the following updated one:
`prepare specific block device, export built image onto it and turns it into an external cache source device(for otaproxy)`.

File: tools/offline_ota_image_builder/README.md (hunk 2 above)
Comment (Bodong-Yang) [reply tools/offline_ota_image_builder/README.md:13]: Emmm yes currently all documents are hosted on confluence. I will temporarily remove the link here.

File: tools/offline_ota_image_builder/README.md (hunk 3 above)
Comment (Bodong-Yang) [reply tools/offline_ota_image_builder/README.md:18]: Ah what I mean here is this tool expecting the input OTA images has the same image layout with the image generated by `ota-metadata`.

What about the following rephrased sentence:
This package currently only recognizes the OTA image generated by [ota-metadata@a711013](https://github.com/tier4/ota-metadata/commit/a711013) script, or the image layout compatible with the `ota-metadata` specification.

Comment (obi-t4) [reply tools/offline_ota_image_builder/README.md:18]: > What about the following rephrased sentence:

Good!

File: tools/offline_ota_image_builder/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (obi-t4) [reply tools/offline_ota_image_builder/README.md:3]: It seems it is difficult to understand the difference `several OTA images` and `one offline OTA image.` in this sentence.

What about this sentence?
```
tools.offline_ota_image_builder is a helper package for bundling offline OTA image from the OTA images build by ota-metadata.
```

File: tools/offline_ota_image_builder/builder.py (hunk 9 above)
Comment (Bodong-Yang) [reply tools/offline_ota_image_builder/builder.py:51]: @obi-t4 python tar stdlib is very slow as it is written in pure python.

File: tools/offline_ota_image_builder/__main__.py (hunk 7 above)
Comment (Bodong-Yang) [reply tools/offline_ota_image_builder/__main__.py]: Updated in [95a3341](https://github.com/tier4/ota-client/pull/252/commits/95a3341da83a815ebe4eeb6c2240d3d71a7d87c8).
Now this package will exit with `EINVAL` on invalid parameter specified.

File: tools/offline_ota_image_builder/configs.py (hunk 11 above)
Comment (Bodong-Yang) [reply tools/offline_ota_image_builder/configs.py]: Updated in [085de7a](https://github.com/tier4/ota-client/pull/252/commits/085de7afcdfc8ce89152b0af42c4071b3a55f8f5).
Now the builder will use the methods and classes defined in `otaclient.app.ota_metadata` to parse the metadata.jwt and processing the image with the parsed metadata.jwt.

File: tools/offline_ota_image_builder/configs.py (hunk 10 above)
Comment (Bodong-Yang) [reply tools/offline_ota_image_builder/configs.py]: Updated in [085de7a](https://github.com/tier4/ota-client/pull/252/commits/085de7afcdfc8ce89152b0af42c4071b3a55f8f5).
Now the builder will use the methods and classes defined in otaclient.app.ota_metadata to parse the metadata.jwt and processing the image with the parsed metadata.jwt.

File: tools/offline_ota_image_builder/README.md [hunk 13]
```diff
@@ -0,0 +1,175 @@
+# Offline OTA image builder
+
+`tools.offline_ota_image_builder` is a helper package for bundling offline OTA image from the OTA images build by ota-metadata(or images compatible with OTA image specification).
+Built image can be used as external cache source for otaproxy, or used by full offline OTA helper util to trigger fully offline OTA update.
+
+**NOTE**: full offline OTA helper util is not yet implemented.
+
+It provides the following features:
+
+1. build offline OTA image from several OTA images,
```
Comment (obi-t4) [inline tools/offline_ota_image_builder/README.md:10]: I'd like to fix `several OTA images` like this:
```suggestion
1. build offline OTA image from the OTA images build by ota-metadata(or images compatible with OTA image specification)
```

**Review by obi-t4 — COMMENTED** [review]
I left a comment. Basically it is OK.

File: tools/offline_ota_image_builder/README.md (hunk 13 above)
Comment (Bodong-Yang) [reply tools/offline_ota_image_builder/README.md]: Updated in [9e61c5e](https://github.com/tier4/ota-client/pull/252/commits/9e61c5e069f7192b4e800e9f8b23b31253414502).

**Review by obi-t4 — APPROVED** [review]
LGTM :+1:

**Pull request merged**